lib/internal/modules/run_main.js

```diff
@@ -92,10 +93,24 @@ function shouldUseESMLoader(mainPath) {
  */
 function runMainESM(mainPath) {
   const { loadESM } = require('internal/process/esm_loader');
-  const { pathToFileURL } = require('internal/url');
+  const { pathToFileURL, URL } = require('internal/url');
+  const _importMapPath = getOptionValue('--experimental-import-map');
   const main = pathToFileURL(mainPath).href;
 
   handleMainPromise(loadESM((esmLoader) => {
+    // Load import map and throw validation errors
+    if (_importMapPath) {
+      const { ImportMap } = require('internal/modules/esm/import_map');
+      const { getCWDURL } = require('internal/util');
+
+      const importMapPath = esmLoader.resolve(_importMapPath, getCWDURL(), { __proto__: null, type: 'json' });
+      return esmLoader.import(importMapPath.url, getCWDURL(), { __proto__: null, type: 'json' })
+        .then((importedMapFile) => {
+          esmLoader.importMap = new ImportMap(importedMapFile.default, new URL(importMapPath.url));
+          return esmLoader.import(main, undefined, { __proto__: null });
+        });
```

Comment:
```suggestion
      return PromisePrototypeThen(esmLoader.import(importMapPath.url, getCWDURL(), { __proto__: null, type: 'json' }), 
        (importedMapFile) => {
          esmLoader.importMap = new ImportMap(importedMapFile.default, new URL(importMapPath.url));
          return esmLoader.import(main, undefined, { __proto__: null });
        });
```
alternatively, await the promise (and make the callback async):
```suggestion
      const importedMapFile = await esmLoader.import(importMapPath.url, getCWDURL(), { __proto__: null, type: 'json' });
      esmLoader.importMap = new ImportMap(importedMapFile.default, new URL(importMapPath.url));
      return esmLoader.import(main, undefined, { __proto__: null });
```

Reply:
+1 for using async/await, you can avoid duplicating the `return` statement if you do. Please see https://github.com/nodejs/node/pull/50590#discussion_r1384614173 for another suggestion that includes this one.